luigi/scheduler.py

```diff
@@ -234,6 +234,27 @@ def load(self):
         else:
             logger.info("No prior state file exists at %s. Starting with clean slate", self._state_path)
 
+    def cascade_disabled_state(self, tasks):
+
+        """ Iterate through all pending tasks, setting the status to DISABLED
+        if at least one of it's dependencies is DISABLED and the rest are
+        either DONE or DISABLED. """
+
+        for task in tasks:
+            if task.status != PENDING:
+                continue
+
+            any_disabled = False
+            should_disable = True
+            for dep in task.deps:
+                t = self.get_task(dep)
+                if t.status == DISABLED:
+                    any_disabled = True
+                should_disable &= (t.status == DISABLED or t.status == DONE)
```

Comment:
A task should be disabled if both of the following are true:

1. At least one of its dependencies are DISABLED (this corresponds to `any_disabled), and
2. All of its dependencies have finished processing. This means all of its dependencies must either be DONE or DISABLED (because FAILED tasks will be rescheduled until they have been disabled by the scheduler). This corresponds to `should_disable`.

So this code sets `any_disabled` to True if any of the dependencies of a task have been DISABLED, and then sets `should_disable` to True only if every dependency is either `DISABLED` or `DONE`